// @vapi-ai/web/dist/vapi.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const daily_js_1 = __importDefault(require("@daily-co/daily-js"));
const events_1 = __importDefault(require("events"));
const client_1 = require("./client");
function destroyAudioPlayer(participantId) {
    const player = document.querySelector(`audio[data-participant-id="${participantId}"]`);
    player?.remove();
}
async function startPlayer(player, track) {
    player.muted = false;
    player.autoplay = true;
    if (track != null) {
        player.srcObject = new MediaStream([track]);
        await player.play();
    }
}
async function buildAudioPlayer(track, participantId) {
    const player = document.createElement('audio');
    player.dataset.participantId = participantId;
    document.body.appendChild(player);
    await startPlayer(player, track);
    return player;
}
function subscribeToTracks(e, call) {
    if (e.participant.local)
        return;
    call.updateParticipant(e.participant.session_id, {
        setSubscribedTracks: {
            audio: true,
            video: false,
        },
    });
}
class VapiEventEmitter extends events_1.default {
    on(event, listener) {
        super.on(event, listener);
        return this;
    }
    once(event, listener) {
        super.once(event, listener);
        return this;
    }
    emit(event, ...args) {
        return super.emit(event, ...args);
    }
    removeListener(event, listener) {
        super.removeListener(event, listener);
        return this;
    }
    removeAllListeners(event) {
        super.removeAllListeners(event);
        return this;
    }
}
class Vapi extends VapiEventEmitter {
    started = false;
    call = null;
    speakingTimeout = null;
    averageSpeechLevel = 0;
    constructor(apiToken, apiBaseUrl) {
        super();
        client_1.client.baseUrl = apiBaseUrl ?? 'https://api.vapi.ai';
        client_1.client.setSecurityData(apiToken);
    }
    cleanup() {
        this.started = false;
        this.call?.destroy();
        this.call = null;
        this.speakingTimeout = null;
    }
    async start(assistant, assistantOverrides) {
        if (this.started) {
            return null;
        }
        this.started = true;
        try {
            const webCall = (await client_1.client.call.callControllerCreateWebCall({
                assistant: typeof assistant === 'string' ? undefined : assistant,
                assistantId: typeof assistant === 'string' ? assistant : undefined,
                assistantOverrides,
            })).data;
            if (this.call) {
                this.cleanup();
            }
            this.call = daily_js_1.default.createCallObject({
                audioSource: true,
                videoSource: false,
            });
            this.call.iframe()?.style.setProperty('display', 'none');
            this.call.on('left-meeting', () => {
                this.emit('call-end');
                this.cleanup();
            });
            this.call.on('participant-left', (e) => {
                if (!e)
                    return;
                destroyAudioPlayer(e.participant.session_id);
            });
            this.call.on('error', () => {
                // Ignore error
            });
            this.call.on('track-started', async (e) => {
                if (!e || !e.participant)
                    return;
                if (e.participant?.local)
                    return;
                if (e.track.kind !== 'audio')
                    return;
                await buildAudioPlayer(e.track, e.participant.session_id);
                if (e?.participant?.user_name !== 'Vapi Speaker')
                    return;
                this.call?.sendAppMessage('playable');
            });
            this.call.on('participant-joined', (e) => {
                if (!e || !this.call)
                    return;
                subscribeToTracks(e, this.call);
            });
            await this.call.join({
                url: webCall.webCallUrl,
                subscribeToTracksAutomatically: false,
            });
            this.call.startRemoteParticipantsAudioLevelObserver(100);
            this.call.on('remote-participants-audio-level', (e) => {
                if (e)
                    this.handleRemoteParticipantsAudioLevel(e);
            });
            this.call.on('app-message', (e) => this.onAppMessage(e));
            this.call.updateInputSettings({
                audio: {
                    processor: {
                        type: 'noise-cancellation',
                    },
                },
            });
            return webCall;
        }
        catch (e) {
            console.error(e);
            this.emit('error', e);
            this.cleanup();
            return null;
        }
    }
    onAppMessage(e) {
        if (!e)
            return;
        try {
            if (e.data === 'listening') {
                return this.emit('call-start');
            }
            else {
                try {
                    const parsedMessage = JSON.parse(e.data);
                    this.emit('message', parsedMessage);
                }
                catch (parseError) {
                    console.log('Error parsing message data: ', parseError);
                }
            }
        }
        catch (e) {
            console.error(e);
        }
    }
    handleRemoteParticipantsAudioLevel(e) {
        const speechLevel = Object.values(e.participantsAudioLevel).reduce((a, b) => a + b, 0);
        this.emit('volume-level', Math.min(1, speechLevel / 0.15));
        const isSpeaking = speechLevel > 0.01;
        if (!isSpeaking)
            return;
        if (this.speakingTimeout) {
            clearTimeout(this.speakingTimeout);
            this.speakingTimeout = null;
        }
        else {
            this.emit('speech-start');
        }
        this.speakingTimeout = setTimeout(() => {
            this.emit('speech-end');
            this.speakingTimeout = null;
        }, 1000);
    }
    stop() {
        this.started = false;
        this.call?.destroy();
        this.call = null;
    }
    send(message) {
        this.call?.sendAppMessage(JSON.stringify(message));
    }
    setMuted(mute) {
        try {
            if (!this.call) {
                throw new Error('Call object is not available.');
            }
            this.call.setLocalAudio(!mute);
        }
        catch (error) {
            throw error;
        }
    }
    isMuted() {
        try {
            if (!this.call) {
                return false;
            }
            return this.call.localAudio() === false;
        }
        catch (error) {
            throw error;
        }
    }
}
exports.default = Vapi;
